diff --git a/src/mdformat/_conf.py b/src/mdformat/_conf.py
index b42f008..11322b3 100644
--- a/src/mdformat/_conf.py
+++ b/src/mdformat/_conf.py
@@ -7,7 +7,7 @@ from typing import Mapping
 from mdformat._compat import tomllib
 
 DEFAULT_OPTS = {
-    "wrap": "keep",
+    "wrap": 80,
     "number": False,
     "end_of_line": "lf",
 }
diff --git a/src/mdformat/renderer/_context.py b/src/mdformat/renderer/_context.py
index 59ce666..da790ff 100644
--- a/src/mdformat/renderer/_context.py
+++ b/src/mdformat/renderer/_context.py
@@ -113,13 +113,6 @@ def text(node: RenderTreeNode, context: RenderContext) -> str:
     # This escape has to be first, else we start multiplying backslashes.
     text = text.replace("\\", "\\\\")
 
-    text = escape_asterisk_emphasis(text)  # Escape emphasis/strong marker.
-    text = escape_underscore_emphasis(text)  # Escape emphasis/strong marker.
-    text = text.replace("[", "\\[")  # Escape link label enclosure
-    text = text.replace("]", "\\]")  # Escape link label enclosure
-    text = text.replace("<", "\\<")  # Escape URI enclosure
-    text = text.replace("`", "\\`")  # Escape code span marker
-
     # Escape "&" if it starts a sequence that can be interpreted as
     # a character reference.
     text = RE_CHAR_REFERENCE.sub(r"\\\g<0>", text)
diff --git a/src/mdformat/renderer/_util.py b/src/mdformat/renderer/_util.py
index 2b3f5c3..2b03766 100644
--- a/src/mdformat/renderer/_util.py
+++ b/src/mdformat/renderer/_util.py
@@ -75,7 +75,7 @@ def maybe_add_link_brackets(link: str) -> str:
 def get_list_marker_type(node: RenderTreeNode) -> str:
     if node.type == "bullet_list":
         mode = "bullet"
-        primary_marker = "-"
+        primary_marker = "*"
         secondary_marker = "*"
     else:
         mode = "ordered"
