diff --git a/build.sh b/build.sh
index e6d0d096f..04cece8fc 100755
--- a/build.sh
+++ b/build.sh
@@ -595,14 +595,8 @@ mkdir -p $BUILD
 . $(dirname $(which $0))/version.sh
 
 # qmake needs absolute paths, so we get them now:
-#   OSX does not have `readlink -f` command. Use equivalent Perl script.
-if [ "$IS_MAC" = "no" ]; then
-  BUILD=`readlink -f $BUILD`
-  BIN=`readlink -f $BIN`
-else
-  BUILD=`perl -MCwd -le 'print Cwd::abs_path(shift)' $BUILD`
-  BIN=`perl -MCwd -le 'print Cwd::abs_path(shift)' $BIN`
-fi
+BUILD=`$PYTHON -c "import os; print(os.path.abspath('$BUILD'), end='')"`
+BIN=`$PYTHON -c "import os; print(os.path.abspath('$BIN'), end='')"`
 
 if [ "$IS_MAC" = "no" ]; then
   if ( gmake -v >/dev/null 2>/dev/null ); then
