diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4388b50d..b23ccdbf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -57,6 +57,8 @@ if( APPLE )
     set(make_env ${CMAKE_COMMAND} -E env SDKROOT=${CMAKE_OSX_SYSROOT})
 endif()
 
+option(USE_SYSTEM_ZLIB "Use system shared ZLIB" OFF)
+
 # run make to extract compiler options, linker options and list of source files
 execute_process(
   COMMAND
@@ -70,6 +72,7 @@ execute_process(
         CC=${CMAKE_C_COMPILER}
         CXX=${CMAKE_CXX_COMPILER}
         LD=${CMAKE_CXX_COMPILER}
+        USE_SYSTEM_ZLIB=${USE_SYSTEM_ZLIB}
         cmake_info
   WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   OUTPUT_VARIABLE MAKE_OUTPUT
@@ -94,10 +97,16 @@ if(ABC_USE_NAMESPACE)
     set_source_files_properties(${ABC_SRC} PROPERTIES LANGUAGE CXX)
 endif()
 
+
+find_package(ZLIB)
+
 function(abc_properties target visibility)
     target_include_directories(${target} ${visibility} ${CMAKE_CURRENT_SOURCE_DIR}/src )
     target_compile_options_filtered(${target} ${visibility} ${ABC_CFLAGS} ${ABC_CXXFLAGS} -Wno-unused-but-set-variable )
     target_link_libraries(${target} ${visibility} ${ABC_LIBS})
+    if(USE_SYSTEM_ZLIB)
+        target_link_libraries(${target} ${visibility} ZLIB::ZLIB)
+    endif()
 endfunction()
 
 set(ABC_MAIN_SRC src/base/main/main.c)
diff --git a/Makefile b/Makefile
index 556c5f0e..e9a46d27 100644
--- a/Makefile
+++ b/Makefile
@@ -15,6 +15,8 @@ $(info $(MSG_PREFIX)Using LD=$(LD))
 PROG := abc
 OS := $(shell uname -s)
 
+USE_SYSTEM_ZLIB ?= OFF
+
 MODULES := \
 	$(wildcard src/ext*) \
 	src/base/abc src/base/abci src/base/cmd src/base/io src/base/main src/base/exor \
@@ -22,7 +24,7 @@ MODULES := \
 	src/map/mapper src/map/mio src/map/super src/map/if \
 	src/map/amap src/map/cov src/map/scl src/map/mpm \
 	src/misc/extra src/misc/mvc src/misc/st src/misc/util src/misc/nm \
-	src/misc/vec src/misc/hash src/misc/tim src/misc/bzlib src/misc/zlib \
+	src/misc/vec src/misc/hash src/misc/tim src/misc/bzlib \
 	src/misc/mem src/misc/bar src/misc/bbl src/misc/parse \
 	src/opt/cut src/opt/fxu src/opt/fxch src/opt/rwr src/opt/mfs src/opt/sim \
 	src/opt/ret src/opt/fret src/opt/res src/opt/lpk src/opt/nwk src/opt/rwt \
@@ -35,6 +37,10 @@ MODULES := \
 	src/aig/aig src/aig/saig src/aig/gia src/aig/ioa src/aig/ivy src/aig/hop \
 	src/aig/miniaig
 
+ifneq (${USE_SYSTEM_ZLIB},ON)
+MODULES += src/misc/zlib
+endif
+
 all: $(PROG)
 default: $(PROG)
 
