diff --git a/src/rcx/src/OpenRCX.tcl b/src/rcx/src/OpenRCX.tcl
index 1e9f0afe5..a522533c3 100644
--- a/src/rcx/src/OpenRCX.tcl
+++ b/src/rcx/src/OpenRCX.tcl
@@ -72,7 +72,8 @@ proc extract_parasitics { args } {
         -debug_net_id
         -context_depth
         -cc_model } \
-      flags { -lef_res }
+      flags { -lef_res
+              -no_merge_via_res }
 
   set ext_model_file ""
   if { [info exists keys(-ext_model_file)] } {
