diff --git a/src/drt/src/dr/FlexGridGraph_maze.cpp b/src/drt/src/dr/FlexGridGraph_maze.cpp
index b3298cc9a..0da5e2322 100644
--- a/src/drt/src/dr/FlexGridGraph_maze.cpp
+++ b/src/drt/src/dr/FlexGridGraph_maze.cpp
@@ -58,7 +58,7 @@ void FlexGridGraph::printExpansion(const FlexWavefrontGrid& currGrid,
       pt,
       currGrid.getCost(),
       currGrid.getPathCost(),
-      currGrid.getLastDir(),
+      int(currGrid.getLastDir()),
       currGrid.getCost() - currGrid.getPathCost(),
       gridX,
       gridY);
diff --git a/src/gui/src/tclCmdInputWidget.h b/src/gui/src/tclCmdInputWidget.h
index 86372b22c..94a89601d 100644
--- a/src/gui/src/tclCmdInputWidget.h
+++ b/src/gui/src/tclCmdInputWidget.h
@@ -33,6 +33,10 @@
 #pragma once
 
 #include <tcl.h>
+// Workaround for https://github.com/swig/swig/issues/2887
+#ifndef TCL_SIZE_MAX
+using Tcl_Size = int;
+#endif
 
 #include <QCompleter>
 #include <QMenu>
