diff --git a/scripts/configure b/scripts/configure
index 46b1549..5cfee2d 100755
--- a/scripts/configure
+++ b/scripts/configure
@@ -5383,9 +5383,7 @@ fi
 
 
 
-if test "x${CSH}" = "xno"; then
-  as_fn_error $? "cannot find /bin/csh---cannot compile!" "$LINENO" 5
-fi
+
 
 # Extract the first word of "python3", so it can be a program name with args.
 set dummy python3; ac_word=$2
diff --git a/scripts/makedbh b/scripts/makedbh
index 01e4fa5..50426de 100755
--- a/scripts/makedbh
+++ b/scripts/makedbh
@@ -1,4 +1,4 @@
-#!/bin/csh -f
+#!/usr/bin/env tcsh -f
 #
 # makes the "database.h" (1st argument, $1) file from "database.h.in"
 # (2nd argument, $2), setting various mask operation definitions
diff --git a/scripts/printmans b/scripts/printmans
index 05e2e1f..8d63c03 100755
--- a/scripts/printmans
+++ b/scripts/printmans
@@ -1,4 +1,4 @@
-#!/bin/csh -f
+#!/usr/bin/env tcsh -f
 
 set cmd=($1)
 shift argv
